internal/provider/resource_secret_access.go

```diff
@@ -0,0 +1,332 @@
+		addClientNotConfiguredError(&resp.Diagnostics, "secret_access", "read")
+		return
+	}
+
+	var state secretAccessResourceModel
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	readSecretOutput, err := s.client.Secrets.ReadSecret(&juju.ReadSecretInput{
+		SecretId:  state.SecretId.ValueString(),
+		ModelName: state.Model.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read secret, got error: %s", err))
+		return
+	}
+
+	// Save the secret details into the Terraform state
+	secretApplications, errDiag := types.ListValueFrom(ctx, types.StringType, readSecretOutput.Applications)
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Applications = secretApplications
+
+	// Save state into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("read secret access %q", state.SecretId))
+}
+
+func (s *secretAccessResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret_access", "update")
+		return
+	}
+
+	var plan, state secretAccessResourceModel
+
+	// Read Terraform plan and state into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var err error
+	noChange := true
+
+	var updatedSecretAccessInput juju.GrantRevokeSecretAccessInput
+
+	updatedSecretAccessInput.ModelName = state.Model.ValueString()
+	updatedSecretAccessInput.SecretId = state.SecretId.ValueString()
+
+	var applicationsInPlanNotInState, applicationsInStateNotInPlan interface{}
+	if !plan.Applications.Equal(state.Applications) {
+		noChange = false
+
+		applicationsInPlanNotInState = difference(plan.Applications, state.Applications)
+		applicationsInStateNotInPlan = difference(state.Applications, plan.Applications)
+
+		resp.Diagnostics.Append(plan.Applications.ElementsAs(ctx, &state.Applications, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		resp.Diagnostics.Append(plan.Applications.ElementsAs(ctx, &updatedSecretAccessInput.Applications, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	if noChange {
+		return
+	}
+
+	// revoke access to applications that are in the state but not in the plan
+	if applicationsInStateNotInPlan != nil {
+		err = s.client.Secrets.RevokeSecretAccess(&juju.GrantRevokeSecretAccessInput{
+			ModelName:    state.Model.ValueString(),
+			SecretId:     state.SecretId.ValueString(),
+			Applications: applicationsInStateNotInPlan.([]string),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Failed to revoke secret access", err.Error())
+			return
+		}
```

Comment:
`s.trace(fmt.Sprintf("revoked secret access for applications %v", ........))` might be helpful here for future debugging.